mantle/harness/harness.go

```diff
@@ -437,19 +445,42 @@ func tRunner(t *H, fn func(t *H)) {
 		t.signal <- true
 	}()
 
+	timeoutOccurred := make(chan bool)
+	wrapper := func() {
+		t.timer = time.AfterFunc(t.timeout, func() {
+			if !t.hasSub {
+				timeoutOccurred <- true
+			}
+		})
+		defer t.timer.Stop()
+		// We will defer this line in case runtime.Goexit is called
+		// If this line were to be skipped, deadlock may occur
+		defer func() { timeoutOccurred <- false }()
+		fn(t)
+		t.finished = true
+	}
```

Comment:
After a timeout occurs, the goroutine running `fn(t)` is uninterrupted. It continues running in the background and the goroutine with `tRunner` will unblock and continue. Not sure how we can stop execution of `fn(t)` and clean up the goroutine here.

Reply:
Cancellation with goroutines is usually done via a `Context` object https://www.sohamkamani.com/golang/context-cancellation-and-values/ - other answers at e.g. https://stackoverflow.com/questions/6807590/how-to-stop-a-goroutine

(I also have an old blog about this here https://blog.verbum.org/2013/12/03/cancelling-computation-gcancellable-or-sigint-versus-threads-versus-exceptions/ )

Reply:
Thanks for the read Colin! From what I understand, if I were to use a `Context` object and pass it to a function that I may want to cancel early (such as `fn`) , I would have have to implement non-blocking checks (that listen to the channel returned by `Context.Done()`) within that function.

If this is the case then adding in such checks will be hard because `fn` actually calls the specific run function for each test. Each test has a different run function since each test tests something different.

Reply:
Right yeah, this is tricky. Two ideas come to mind:
- In practice, the things that will hang will be I/O with outside resources. Unless the test code is really broken (and e.g. has an infinite loop of pure computation), it'll be doing some kind of I/O. The major bits I can think of are: platform-specific API requests during provisioning and teardown, and SSH commands. So what we could do is change the model for those critical places to be "cancellable" by supporting a stop signal/`ctx.Done()`. I think just doing the SSH bit will cover a large portion of timeouts (including those provisioning hangs which are actually from kola waiting for SSH access so it can start recording the journal before even running the test's `Run` function).
- A more nuclear option, though it'd require some major rework, is to fork off the tests into separate processes. Then it's simply a matter of killing the child process after the timeout. The caveat there is that there's a risk of leaking, though `PDEATHSIG` helps with that. I wouldn't do this before trying the first idea.

Reply:
@bgilbert might have some other ideas here.

Reply:
Option 1 (change model for API requests and SSH commands to be "cancellable") makes sense to me, I can try that.